application/src/main/java/org/togetherjava/tjbot/commands/system/CommandSystem.java

```diff
@@ -0,0 +1,166 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.togetherjava.tjbot.commands.CommandRegistry;
+import org.togetherjava.tjbot.commands.SlashCommand;
+import org.togetherjava.tjbot.db.Database;
+import org.togetherjava.tjbot.util.TriConsumer;
+
+import java.util.*;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * The command system is the core of command handling in this application.
+ * <p>
+ * It knows and manages all commands, registers them towards Discord and is the entry point of all
+ * events. It forwards events to their corresponding commands and does the heavy lifting on all sort
+ * of event parsing.
+ * <p>
+ * <p>
+ * Commands are made available via {@link CommandRegistry}, then the system has to be added to JDA
+ * as an event listener, using {@link net.dv8tion.jda.api.JDA#addEventListener(Object...)}.
+ * Afterwards, the system is ready and will correctly forward events to all commands.
+ */
+public final class CommandSystem extends ListenerAdapter implements SlashCommandProvider {
+    private static final Logger logger = LoggerFactory.getLogger(CommandSystem.class);
+
+    private final Map<String, SlashCommand> nameToSlashCommands;
+    private static final String RELOAD_COMMAND = "reload";
+
+    /**
+     * Creates a new command system which uses the given database to allow commands to persist data.
+     * <p>
+     * Commands are fetched from {@link CommandRegistry}.
+     *
+     * @param database the database that commands may use to persist data
+     */
+    @SuppressWarnings("ThisEscapedInObjectConstruction")
+    public CommandSystem(@NotNull Database database) {
+        nameToSlashCommands = CommandRegistry.createSlashCommands(database)
+            .stream()
+            .collect(Collectors.toMap(SlashCommand::getName, Function.identity()));
+
+        if (nameToSlashCommands.containsKey(RELOAD_COMMAND)) {
+            throw new IllegalStateException(
+                    "The 'reload' command is a special reserved command that must not be used by other commands");
+        }
+        nameToSlashCommands.put(RELOAD_COMMAND, new ReloadCommand(this));
+
+        if (logger.isInfoEnabled()) {
+            logger.info("Available commands: {}", nameToSlashCommands.values());
+        }
+    }
+
+    @Override
+    public @NotNull Collection<SlashCommand> getSlashCommands() {
+        return Collections.unmodifiableCollection(nameToSlashCommands.values());
+    }
+
+    @Override
+    public @NotNull Optional<SlashCommand> getSlashCommand(@NotNull String name) {
+        return Optional.ofNullable(nameToSlashCommands.get(name));
+    }
+
+    @Override
+    public void onReady(@NotNull ReadyEvent event) {
+        // Register reload on all guilds
+        logger.debug("JDA is ready, registering reload command");
+        event.getJDA().getGuildCache().forEach(this::registerReloadCommand);
+        logger.debug("Command system is now ready");
```

Comment:
I thought about that as well but adding a blocking call there just for the debug-log to be accurate seemed odd to me.

That said, it might actually be a good idea to block the system until its ready anyways. I will look into this.